### PR TITLE
Runtime Upgrades & Storage Migrations

### LINKED ISSUE #580: Better Runtime Upgrade Documentation
We can create more [documentation around runtime upgrades](https://github.com/substrate-developer-hub/substrate-developer-hub.github.io/blob/source/docs/knowledgebase/advanced/executor.md#runtime-upgrades), which is one of Substrate's most important and distinguishing characteristics. In particular, a community member recently asked for an enumeration of the changes that _can_ be included in a  forkless runtime upgrade and those that _cannot_.

### PR DESCRIPTION
**Pull request by danforbes**
Closes #580

### CODE REVIEW

File: docs/knowledgebase/runtime/upgrades.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+FRAME storage migrations are implemented by way of
+[the `OnRuntimeUpgrade` trait](https://substrate.dev/rustdocs/v2.0.0/frame_support/traits/trait.OnRuntimeUpgrade.html),
+which specifies a single function, `on_runtime_upgrade`. This function provides a hook that allows
+runtime developers to specify logic that will run immediately _after_ a runtime upgrade but _before_
+any [extrinsics or even the `on_initialize` function](execution#executing-a-block) have executed.
+
+### Preparing for a Migration
+
+Preparing for a storage migrations means understanding the changes that are defined by a runtime
```
Comment (apopiak) [inline docs/knowledgebase/runtime/upgrades.md:43]:
```suggestion
Preparing for a storage migration means understanding the changes that are defined by a runtime
```

File: docs/knowledgebase/runtime/upgrades.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+in which the pallets appear in the `construct_runtime!` macro - in particular, they will run in
+_reverse_ (top-to-bottom) order. FRAME exposes a capability to inject storage migrations in a custom
+order, if needed (see an
+[example here](https://github.com/hicommonwealth/edgeware-node/blob/7b66f4f0a9ec184fdebcccd41533acc728ebe9dc/node/runtime/src/lib.rs#L845-L866)).
+
+FRAME storage migrations will run in this order:
+
+1. `frame_system::on_runtime_upgrade`
+1. Custom `on_runtime_upgrade`, as described above
+1. Any pallet's `on_runtime_upgrade` function, in the order described above
```
Comment (apopiak) [inline docs/knowledgebase/runtime/upgrades.md:73]:
```suggestion
1. `frame_system::on_runtime_upgrade`
2. Custom `on_runtime_upgrade`, as described above
3. All `on_runtime_upgrade` functions defined in the pallets included in the runtime, in the order described above
```

**Review by apopiak — COMMENTED** [review]
Looking good.
Includes most of the points from https://hackmd.io/@apopiak/Hyhroh0Xv/edit, but not aimed to be a (complete) runtime migration guide (i.e. fill the same role), though, right?

**Review by wheresaddie — APPROVED** [review]

**Review by wheresaddie — COMMENTED** [review]
see my PRs for edits/reworks attached to this pull

**Comment by danforbes** [comment]
> ...not aimed to be a (complete) runtime migration guide (i.e. fill the same role), though, right?

@apopiak I would say this is correct. It may be cool to create a storage migration recipe or maybe an advanced runtime upgrade tutorial. Do you think there are any major conceptual gaps in this document?

**Review by apopiak — APPROVED** [review]

**Pull request merged**